pkg/datapath/linux/ipsec/xfrm_collector.go

```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2022 Authors of Cilium
+
+package ipsec
+
+import (
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/procfs"
+
+	"github.com/cilium/cilium/pkg/metrics"
+)
+
+const (
+	labelErrorType         = "type"
+	labelErrorTypeInbound  = "inbound"
+	labelErrorTypeOutbound = "inbound"
+)
+
+type xfrmCollector struct {
+	xfrmStatFunc func() (procfs.XfrmStat, error)
+
+	// Inbound errors
+	xfrmInErrorDesc           *prometheus.Desc
+	xfrmInBufferErrorDesc     *prometheus.Desc
+	xfrmInHdrErrorDesc        *prometheus.Desc
+	xfrmInNoStatesDesc        *prometheus.Desc
+	xfrmInStateProtoErrorDesc *prometheus.Desc
+	xfrmInStateModeErrorDesc  *prometheus.Desc
+	xfrmInStateSeqErrorDesc   *prometheus.Desc
+	xfrmInStateExpiredDesc    *prometheus.Desc
+	xfrmInStateMismatchDesc   *prometheus.Desc
+	xfrmInStateInvalidDesc    *prometheus.Desc
+	xfrmInTmplMismatchDesc    *prometheus.Desc
+	xfrmInNoPolsDesc          *prometheus.Desc
+	xfrmInPolBlockDesc        *prometheus.Desc
+	xfrmInPolErrorDesc        *prometheus.Desc
+	xfrmAcquireErrorDesc      *prometheus.Desc
+	xfrmFwdHdrErrorDesc       *prometheus.Desc
+
+	// Outbound errors
+	xfrmOutErrorDesc            *prometheus.Desc
+	xfrmOutBundleGenErrorDesc   *prometheus.Desc
+	xfrmOutBundleCheckErrorDesc *prometheus.Desc
+	xfrmOutNoStatesDesc         *prometheus.Desc
+	xfrmOutStateProtoErrorDesc  *prometheus.Desc
+	xfrmOutStateModeErrorDesc   *prometheus.Desc
+	xfrmOutStateSeqErrorDesc    *prometheus.Desc
+	xfrmOutStateExpiredDesc     *prometheus.Desc
+	xfrmOutPolBlockDesc         *prometheus.Desc
+	xfrmOutPolDeadDesc          *prometheus.Desc
+	xfrmOutPolErrorDesc         *prometheus.Desc
+	xfrmOutStateInvalidDesc     *prometheus.Desc
+}
+
+// NewXFRMCollector returns a new prometheus.Collector for /proc/net/xfrm_stat
+// https://www.kernel.org/doc/Documentation/networking/xfrm_proc.txt
+func NewXFRMCollector() prometheus.Collector {
+	return newXFRMCollector(procfs.NewXfrmStat)
+}
+
+func newXFRMCollector(statFn func() (procfs.XfrmStat, error)) prometheus.Collector {
+	return &xfrmCollector{
+		xfrmStatFunc: statFn,
+
+		// Inbound errors
+		xfrmInErrorDesc: prometheus.NewDesc(
+			prometheus.BuildFQName(metrics.Namespace, subsystem, "xfrm_in_error"),
+			"All inbound errors which is not matched others",
+			nil, map[string]string{labelErrorType: labelErrorTypeInbound},
```

Comment:
I am thinking about about adding labels for cluster name and node name, but existing ebpf related metrics don't have these label :man_shrugging: